File: finalyearproject/repos/dkd-platform-libs/sdk/typescript/src/inbox.ts
// HAND-AUTHORED platform primitive (NOT dkdgen-generated).
// Consumer inbox — the "I" of SA-CONV-QUARTET (EF §21.1 / EF-EVT-6). A consumer records the
// event_id in the SAME transaction as its side effect, so a redelivered event is a no-op.
// Keyed by (consumer, event_id): the same fact may be consumed independently by many contexts.
// One of five byte-identical language implementations sharing the SAME contract (PL-02).
import type { DbExecutor } from "./outbox.js";

export class Inbox {
  /** True when (consumer, eventId) was already processed — dedup guard, same tx as the effect. */
  static async alreadyProcessed(
    tx: DbExecutor,
    consumer: string,
    eventId: string,
  ): Promise<boolean> {
    const res = await tx.query(
      "SELECT 1 FROM inbox WHERE consumer = $1 AND event_id = $2",
      [consumer, eventId],
    );
    return res.rowCount >= 1;
  }

  /**
   * Record (consumer, eventId) as processed inside the effect's tx. Idempotent under concurrent
   * redelivery (ON CONFLICT DO NOTHING on the composite PK).
   * @returns true when this call claimed the event, false when already marked.
   */
  static async markProcessed(
    tx: DbExecutor,
    consumer: string,
    eventId: string,
  ): Promise<boolean> {
    const res = await tx.query(
      "INSERT INTO inbox(consumer, event_id, processed_at) VALUES ($1,$2,now()) " +
        "ON CONFLICT (consumer, event_id) DO NOTHING",
      [consumer, eventId],
    );
    return res.rowCount === 1;
  }
}

File: finalyearproject/repos/dkd-platform-libs/sdk/go/inbox.go
package dkdplatform

import (
	"context"
	"fmt"
)

// Inbox is the consumer-side dedup half of SA-CONV-QUARTET. AlreadyProcessed and
// MarkProcessed MUST run in the SAME transaction as the side effect they guard,
// so a consumer that crashes after the side effect but before the mark replays
// safely (effectively-once). Stateless; construct with a zero value.
type Inbox struct{}

// AlreadyProcessed reports whether (consumer, eventId) has been recorded. Run it
// inside the handler's transaction, before applying the side effect.
func (Inbox) AlreadyProcessed(ctx context.Context, tx DB, consumer, eventID string) (bool, error) {
	if tx == nil {
		return false, fmt.Errorf("inbox: nil tx handle")
	}
	rows, err := tx.QueryContext(ctx,
		`SELECT 1 FROM inbox WHERE consumer = $1 AND event_id = $2`, consumer, eventID)
	if err != nil {
		return false, fmt.Errorf("inbox: lookup %s/%s: %w", consumer, eventID, err)
	}
	defer rows.Close()
	found := rows.Next()
	if err := rows.Err(); err != nil {
		return false, fmt.Errorf("inbox: lookup %s/%s: %w", consumer, eventID, err)
	}
	return found, nil
}

// MarkProcessed records (consumer, eventId) in the handler's transaction. ON
// CONFLICT DO NOTHING keeps it idempotent under concurrent redelivery.
func (Inbox) MarkProcessed(ctx context.Context, tx DB, consumer, eventID string) error {
	if tx == nil {
		return fmt.Errorf("inbox: nil tx handle")
	}
	if consumer == "" || eventID == "" {
		return fmt.Errorf("inbox: consumer and eventId are required")
	}
	_, err := tx.ExecContext(ctx,
		`INSERT INTO inbox (consumer, event_id, processed_at)
		 VALUES ($1,$2,now())
		 ON CONFLICT (consumer, event_id) DO NOTHING`, consumer, eventID)
	if err != nil {
		return fmt.Errorf("inbox: mark %s/%s: %w", consumer, eventID, err)
	}
	return nil
}
